### PR TITLE
Use 100% DPI for histogram calculation to avoid D2DERR_INTERMEDIATE_TOO_LARGE

### PR DESCRIPTION
**Pull request by imbushuo**
I think it's kind of weird and fun to have this repro on screen beyond 4K, but anyway it works at higher resolution now. Is the issue possible from Direct2D itself?

Untitled

Windows AC Example has the exact same issue, and this fix can be applied there as well.

### CODE REVIEW

**Comment by Filoppi** [comment]
I'm sorry for the stupid question, but how did you get to build this app? I just get libheif include errors (`missing libheif/heif.h`), even if I have its nuget package and even if I manually add them to the project.

**Comment by 13thsymphony** [comment]
@imbushuo thanks for the PR and sorry for the radio silence!

This looks like a righteous bug and fix; I can't personally repro the bug (my GPU is too good??) but I definitely should be using 96 DPI for histogram rendering (and possibly for image rendering too). However I'd like to reasonably ensure that the histogram always returns some reasonable value, versus ignoring texture limit exceeded issues.

I made a new commit which explicitly caps the texture size:
[1483b3c](https://github.com/13thsymphony/HDRImageViewer/commit/1483b3c438a04828cf408bd4206f72e8519d66d4)

If you agree, I'd like to keep your 96.0DPI change, but I don't want to ignore D2DERR_INTERMEDIATE_TOO_LARGE - with our combined changes this should once again become an exceptional situation, as the histogram texture and intermediates should never exceed 2048 pixel size.

File: DXRenderer/HDRImageViewerRenderer.cpp [hunk 1]
```diff
@@ -808,14 +808,23 @@ void HDRImageViewerRenderer::ComputeHdrMetadata()
...
     // We ignore D2DERR_RECREATE_TARGET here. This error indicates that the device
     // is lost. It will be handled during the next call to Present.
+    //
+    // We also ignore D2DERR_INTERMEDIATE_TOO_LARGE in case D2D complains the
+    // intermediate tile being too large, the picture should show properly, only
+    // MaxCLL estimation won't work (show as N/A.)
     HRESULT hr = ctx->EndDraw();
-    if (hr != D2DERR_RECREATE_TARGET)
+    ctx->SetDpi(m_deviceResources->GetDpi(), m_deviceResources->GetDpi());
+    if (hr != D2DERR_RECREATE_TARGET && hr != D2DERR_INTERMEDIATE_TOO_LARGE)
```
Comment (13thsymphony) [inline DXRenderer/HDRImageViewerRenderer.cpp]: Per my comment in the PR can you rebase on my latest change and remove the ignore D2DERR_INTERMEDIATE_TOO_LARGE?
(https://github.com/13thsymphony/HDRImageViewer/commit/1483b3c438a04828cf408bd4206f72e8519d66d4)

File: DXRenderer/HDRImageViewerRenderer.cpp [hunk 2]
```diff
@@ -808,14 +808,23 @@ void HDRImageViewerRenderer::ComputeHdrMetadata()
 
     auto ctx = m_deviceResources->GetD2DDeviceContext();
 
+    // Reset DPI to 100% (96.0f) during histogram calculation, D2D complains
```
Comment (13thsymphony) [inline DXRenderer/HDRImageViewerRenderer.cpp]: Per my PR comment, I believe we no longer will hit the max texture size issue, so can you update this comment to just say that histogram rendering should always occur without DPI scaling (i.e. on raw pixels)?

File: DXRenderer/HDRImageViewerRenderer.cpp (hunk 1 above)
Comment (imbushuo) [reply DXRenderer/HDRImageViewerRenderer.cpp]: Let me take a look later today or maybe this weekend. Just traveled back from Bellevue/Redmond.

By the way, which GPU are you using? Mine is RTX A5000 so technically it falls into the category of some decent card.

Comment (13thsymphony) [reply DXRenderer/HDRImageViewerRenderer.cpp]: Take your time!

I'm using an RTX 2080....so yours (Ampere) beats mine (Turing) :). And this confuses me as to my inability to repro the issue, I had thought that this was a DX feature level or GPU architecture limitation. I was using the repro image from #47 and rendering to a 4K display (which shouldn't matter) at 300% DPI.

**Comment by imbushuo** [comment]
Updated commit & validated 6K images based on comment.

**Review by 13thsymphony — APPROVED** [review]

**Comment by 13thsymphony** [comment]
Thanks for your help! Approved and squash/merged.

**Comment by Filoppi** [comment]
@13thsymphony Maybe I ask if you are planning on making a new Microsoft Store release of the app with this fix?

**Comment by ValZapod** [comment]
> @13thsymphony Maybe I ask if you are planning on making a new Microsoft Store release of the app with this fix?

Yeah?

**Comment by 13thsymphony** [comment]
@Filoppi @ValZapod Sorry for the delay, yes I'm working on getting an updated version to the Store soon.

**Pull request merged**